src/imap.rs

```diff
@@ -542,45 +557,58 @@ impl Imap {
             return 0;
         }
 
-        match self.session.lock().unwrap().0 {
+        let teardown: bool;
+
+        match &mut *self.session.lock().unwrap() {
             Some(ref mut session) => {
                 if let Ok(caps) = session.capabilities() {
-                    let can_idle = caps.has("IDLE");
-                    let has_xlist = caps.has("XLIST");
-
-                    let caps_list = caps.iter().fold(String::new(), |mut s, c| {
-                        s += " ";
-                        s += c;
-                        s
-                    });
+                    if !context.sql.is_open() {
+                        warn!(
+                            context,
+                            0,
+                            "IMAP-LOGIN as {} ok but ABORTING",
+                            as_str(lp.mail_user),
+                        );
+                        teardown = true;
+                    } else {
+                        let can_idle = caps.has("IDLE");
+                        let has_xlist = caps.has("XLIST");
 
-                    log_event!(
-                        context,
-                        Event::IMAP_CONNECTED,
-                        0,
-                        "IMAP-LOGIN as {} ok",
-                        as_str(lp.mail_user),
-                    );
-                    info!(context, 0, "IMAP-capabilities:{}", caps_list);
+                        let caps_list = caps.iter().fold(String::new(), |mut s, c| {
+                            s += " ";
+                            s += c;
+                            s
+                        });
 
-                    let mut config = self.config.write().unwrap();
-                    config.can_idle = can_idle;
-                    config.has_xlist = has_xlist;
-                    *self.connected.lock().unwrap() = true;
+                        log_event!(
+                            context,
+                            Event::IMAP_CONNECTED,
+                            0,
+                            "IMAP-LOGIN as {} ok",
+                            as_str(lp.mail_user),
+                        );
+                        info!(context, 0, "IMAP-capabilities:{}", caps_list);
 
-                    1
+                        self.config.write().unwrap().can_idle = can_idle;
+                        self.config.write().unwrap().has_xlist = has_xlist;
+                        *self.connected.lock().unwrap() = true;
+                        teardown = false;
+                    }
                 } else {
-                    self.unsetup_handle(context);
-                    self.free_connect_params();
-                    0
+                    teardown = true;
                 }
             }
             None => {
-                self.unsetup_handle(context);
-                self.free_connect_params();
-                0
+                teardown = true;
             }
         }
```

Comment:
i think it's important to do the teardown outside the scope of the lock held through ``match &mut *self.session.lock().unwrap()``above because unsetup_handle tries to acquire the lock IIUC.